### PR TITLE
ASoC: SOF: Intel: add 3 more callbacks for HDaudio DMA reuse

### PR DESCRIPTION
**Pull request by plbossart**
The existing code has 3 hard-coded references to HDaudio codecs and links, this will not work if we want to reuse the DMA channel allocation for DMIC/SSP/SoundWire. Add 3 new callbacks and for the HDaudio DAI case use the existing code as is (no functionality change for now)

### CODE REVIEW

File: sound/soc/sof/intel/hda-dai-ops.c
Comment (bardliao) [inline sound/soc/sof/intel/hda-dai-ops.c]: Why do we need `cpu_dai` if it is not used?

File: sound/soc/sof/intel/hda-dai.c
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c]: nitpick:
```c
if (ops->calc_stream_format && ops->setup_hext_stream) {
...
}
```
if ops->setup_hext_stream is not set, no point of calculating in the first place.
Or to modify the setup_hext_stream() callback to to self contain the format_val calculation (modifying the parameter list I guess would be needed).

File: sound/soc/sof/intel/hda.h
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.h:921]: I'm not sure if the `, mainly to program LOSIDV` part is needed or what that actually means.
The retrieved `hlink` is not used (at least for now) for LOSIDV programming, or the comment wants to say that the get_hlink callback should do that?

File: sound/soc/sof/intel/hda.h
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.h]: you probably mean: `codec_dai_set_hdac_stream` or `codec_dai_set_hstream` or just simply `codec_dai_set_stream`?

**Review by ujfalusi — COMMENTED** [review]
If the main intention is to remove the `codec_dai` from the cpu dai code is then I think this just going to make things harder to follow.

I'm not sure what and how the first patch's callback would look like for non HDA mlinks for example.

File: sound/soc/sof/intel/hda-dai-ops.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c]: yes, that's true. I didn't notice it, so far all that's needed in sdev and the substream. This can be simplified, thanks for the suggestion @bardliao

File: sound/soc/sof/intel/hda.h
Comment (plbossart) [reply sound/soc/sof/intel/hda.h:921]: the comment makes no sense, I''ll update this @ujfalusi . Buf so that it's clearer, here's what the callback would look like for SSP:

````c
static struct hdac_ext_link *ssp_get_hlink(struct snd_sof_dev *sdev,
					   struct snd_pcm_substream *substream)
{
	struct hdac_bus *bus = sof_to_bus(sdev);

	return hdac_bus_eml_ssp_get_hlink(bus);
}
````

File: sound/soc/sof/intel/hda.h
Comment (plbossart) [reply sound/soc/sof/intel/hda.h]: yes, this can be simplified, and we can avoid passing the hext_stream when we only need the hdac_stream.

Thanks for the suggestion @ujfalusi

File: sound/soc/sof/intel/hda-dai.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c]: makes sense @ujfalusi I'll do the change

File: sound/soc/sof/intel/hda.h
Comment (plbossart) [reply sound/soc/sof/intel/hda.h]: done

File: sound/soc/sof/intel/hda-dai.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c]: done

File: sound/soc/sof/intel/hda-dai-ops.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c]: removed cpu_dai

**Comment by plbossart** [comment]
v2-changes: 
[v2.diff.txt](https://github.com/thesofproject/linux/files/10805278/v2.diff.txt)

**Comment by plbossart** [comment]
~~hold off on any merges, thsi could be further simplified if https://github.com/thesofproject/linux/pull/4209 works.~~

no dependency, this can be reviewed/merged @RanderWang @bardliao @ujfalusi @ranj063

**Review by bardliao — DISMISSED** [review]

**Review by RanderWang — DISMISSED** [review]

**Review by ujfalusi — DISMISSED** [review]

File: sound/soc/sof/intel/hda-dai.c
Comment (ranj063) [inline sound/soc/sof/intel/hda-dai.c:176]: is this op meant to be optional @plbossart

**Comment by plbossart** [comment]
Conflict resolved, needs re-approval @ranj063 @ujfalusi @RanderWang @ranj063

File: sound/soc/sof/intel/hda-dai.c
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:176]: yes, it's not needed for SoundWire/DMIC/SSP. It's only required for HDaudio codecs.

**Review by ranj063 — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Pull request merged**
